### PR TITLE
Improved RGD fingerprint score

### PR DESCRIPTION
**Pull request by jones-gareth**
#### What does this implement/fix? Explain your changes.

Three improvements to the fingerprint variance scoring function for RGD

1. The current fingerprint comparison for RGD is scored based on the variance for bitstrings.  A penalty is determined for each bit position in Morgan fingerprints. For each R group label we count the number of groups that set each bit position.  The plot here shows the penalty (y) against the bitcount (x) for 10 r groups in a decomposition for any bit position.
image
The issue is is that the penalty for having a bit count of 1 is the same as the penalty for having a bit count of 9, whereas it is clearly much better to have 9 out of a possible 10 bits present, rather than just one. (This is not as awful as initially appears as 9 single bits spread across different bit positions have a much higher penalty that all nine bits in a single position).  To ensure that more bit counts at a bit position always have a lower penalty I transform the counts to give this scoring function (note if no bits are set the penalty is still 0).
image

2. The overall penalty score for a decomposition is the sum of the roots of the group penalties: &sum;(&Sqrt;p<sub>i</sub>).  I have changed this so that the penalty is the root of the sum of the group penalties: &Sqrt;(&sum;p<sub>i</sub>).  The previous formula was creating distortions by encouraging large penalties on single groups rather than spreading the penalties across groups.

3. In decompositions where a group is not present in some decompostions (when the core has a wildcard atom) no fingerprint is used in the current scoring function.  However, we want those positions to align with hydrogen only rgroups.  So, in the new scoring function I give missing rgroups the same fingerprint as a hydrogen.

#### Any other comments?

I have test data that shows all of this stuff improving decompositions.  Unfortunately, the test are built off customer data that I can't share.

I removed `testGeminalRGroups` from `testRGroupInternals`.  This test now fails, but that is fine as we ware testing for selecting against multiple R groups on a single attachment point.  Since that test was written the RGD no longer permits multiple R groups on a single attachment point.

### CODE REVIEW

File: Code/GraphMol/RGroupDecomposition/RGroupFingerprintScore.cpp [hunk 1]
```diff
@@ -24,6 +25,21 @@
 
 namespace RDKit {
 
+namespace detail {
+RData dummyFingerprint(int label) {
```
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupFingerprintScore.cpp:29]: Perhaps a bit more clarity here that a dummy FP is an fp equivalent to hydrogen

Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupFingerprintScore.cpp]: I've renamed the function and added a comment

File: Code/GraphMol/RGroupDecomposition/RGroupFingerprintScore.cpp [hunk 2]
```diff
@@ -190,14 +220,9 @@ double FingerprintVarianceScoreData::fingerprintVarianceGroupScore() {
 #endif
   auto sum = std::accumulate(
       labelsToVarianceData.cbegin(), labelsToVarianceData.cend(), 0.0,
-      [](double sum,
-         std::pair<int, std::shared_ptr<VarianceDataForLabel>> pair) {
+      [this](double sum,
```
[the remark this answers is not in the document]
Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupFingerprintScore.cpp]: Nope. I've taken in out in the latest commit

**Review by bp-kelley — APPROVED** [review]

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
will merge after the CI builds go through

**Pull request merged**